File: src/yoinkc/inspectors/container.py
"""Container inspector: quadlet units, compose files, optional podman query.

Parses Image= from quadlet .container files, image: from compose YAML,
and optionally runs podman inspect for live container details.
"""

import json
import re
from pathlib import Path
from typing import Any, Dict, List, Optional

from ..executor import Executor
from ..schema import (
    ContainerSection, QuadletUnit, ComposeFile, ComposeService,
    RunningContainer, ContainerMount,
)
from .._util import debug as _debug_fn, safe_read as _safe_read_raw, make_warning
from . import filtered_rglob


def _debug(msg: str) -> None:
    _debug_fn("container", msg)


def _safe_glob(d: Path, pattern: str) -> List[Path]:
    try:
        return list(d.glob(pattern))
    except (PermissionError, OSError):
        return []


def _safe_read(p: Path) -> str:
    return _safe_read_raw(p, label="container")


def _extract_quadlet_image(content: str) -> str:
    """Extract the Image= value from a quadlet .container file."""
    for line in content.splitlines():
        stripped = line.strip()
        if stripped.lower().startswith("image") and "=" in stripped:
            key, _, val = stripped.partition("=")
            if key.strip().lower() == "image":
                return val.strip()
    return ""


def _extract_compose_images(content: str) -> List[Dict[str, str]]:
    """Extract image: fields from a compose YAML without requiring PyYAML.

    Returns a list of {service, image} dicts. Uses simple line-by-line
    parsing to avoid adding a dependency.  Detects service-level indent
    dynamically so 2-space, 4-space, and tab-indented files all work.
    """
    results: List[Dict[str, str]] = []
    lines = content.splitlines()
    current_service = ""
    in_services = False
    service_indent: Optional[int] = None  # indent of the first service key seen

    for line in lines:
        stripped = line.strip()
        if not stripped or stripped.startswith("#"):
            continue

        indent = len(line) - len(line.lstrip())

        if stripped == "services:" or stripped.startswith("services:"):
            in_services = True
            service_indent = None
            continue

        # Top-level key that isn't services — exit services block
        if indent == 0 and ":" in stripped and not stripped.startswith("-"):
            key = stripped.split(":")[0].strip()
            if key != "services":
                in_services = False
                service_indent = None
                current_service = ""
                continue

        if not in_services:
            continue

        # Detect service-block indent from the first indented key
        if service_indent is None and indent > 0:
            service_indent = indent

        if service_indent is not None:
            if indent == service_indent and stripped.endswith(":") and not stripped.startswith("-"):
                current_service = stripped.rstrip(":")
                continue

        if current_service:
            m = re.match(r"image:\s*(.+)", stripped)
            if m:
                image_ref = m.group(1).strip().strip("'\"")
                results.append({"service": current_service, "image": image_ref})

    return results


def _parse_podman_inspect(data: List[dict]) -> List[RunningContainer]:
    """Normalize podman inspect JSON into our schema format."""
    results: List[RunningContainer] = []
    for c in data:
        mounts = [
            ContainerMount(
                type=m.get("Type", ""),
                source=m.get("Source", ""),
                destination=m.get("Destination", ""),
                mode=m.get("Mode", ""),
                rw=m.get("RW", True),
            )
            for m in (c.get("Mounts") or [])
        ]

        net_settings = c.get("NetworkSettings") or {}
        networks = {}
        for net_name, net_info in (net_settings.get("Networks") or {}).items():
            networks[net_name] = {
                "ip": net_info.get("IPAddress", ""),
                "gateway": net_info.get("Gateway", ""),
                "mac": net_info.get("MacAddress", ""),
            }

        state = c.get("State") or {}
        results.append(RunningContainer(
            id=c.get("Id", ""),
            name=c.get("Name", ""),
            image=c.get("Image", ""),
            image_id=c.get("ImageID", ""),
            status=state.get("Status", ""),
            mounts=mounts,
            networks=networks,
            ports=net_settings.get("Ports") or {},
            env=(c.get("Config") or {}).get("Env") or [],
        ))
    return results


def run(
    host_root: Path,
    executor: Optional[Executor],
    query_podman: bool = False,
    warnings: Optional[list] = None,
) -> ContainerSection:
    section = ContainerSection()
    host_root = Path(host_root)

    # --- Quadlet units ---
    quadlet_dirs = [
        "etc/containers/systemd",
        "usr/share/containers/systemd",
        "etc/systemd/system",
    ]

    # User-level quadlets: ~/.config/containers/systemd/ for each real user
    passwd = host_root / "etc/passwd"
    try:
        if passwd.exists():
            for line in passwd.read_text().splitlines():
                parts = line.split(":")
                if len(parts) >= 7:
                    try:
                        uid = int(parts[2])
                    except ValueError:
                        continue
                    if 1000 <= uid < 60000:
                        home = parts[5].lstrip("/")
                        quadlet_dirs.append(
                            f"{home}/.config/containers/systemd"
                        )
    except (PermissionError, OSError):
        pass

    for subdir in quadlet_dirs:
        d = host_root / subdir
        try:
            exists = d.exists()
        except (PermissionError, OSError):
            exists = False
        if not exists:
            continue
        for f in _safe_glob(d, "*.container"):
            content = _safe_read(f)
            _debug(f"quadlet: {f} ({len(content)} bytes)")
            image_ref = _extract_quadlet_image(content)
            _debug(f"quadlet: {f.name} Image={image_ref!r}")
            if not image_ref and content:
                _debug(f"quadlet: no Image= found, first 5 lines: {content.splitlines()[:5]}")
            section.quadlet_units.append(QuadletUnit(
                path=str(f.relative_to(host_root)),
                name=f.name,
                content=content,
                image=image_ref,
            ))

    # --- Compose files ---
    for search_dir in ("opt", "srv", "etc"):
        d = host_root / search_dir
        if not d.exists():
            continue
        for pattern in ("docker-compose*.yml", "docker-compose*.yaml",
                        "compose*.yml", "compose*.yaml"):
            for f in filtered_rglob(d, pattern):
                if not f.is_file():
                    continue
                content = _safe_read(f)
                images = [ComposeService(**img) for img in _extract_compose_images(content)]
                section.compose_files.append(ComposeFile(
                    path=str(f.relative_to(host_root)),
                    images=images,
                ))

    # --- Podman query ---
    if query_podman and executor:
        # podman ps for the container list
        r = executor(["podman", "ps", "-a", "--format", "json"])
        if r.returncode != 0 and warnings is not None:
            warnings.append(make_warning(
                "containers",
                "--query-podman requested but podman ps failed — live container data unavailable.",
            ))
        if r.returncode == 0 and r.stdout.strip():
            try:
                ps_data = json.loads(r.stdout)
            except json.JSONDecodeError:
                ps_data = []

            if isinstance(ps_data, list) and ps_data:
                container_ids = [c.get("ID", "") for c in ps_data if c.get("ID")]
                if container_ids:
                    ir = executor(["podman", "inspect"] + container_ids)
                    if ir.returncode == 0 and ir.stdout.strip():
                        try:
                            inspect_data = json.loads(ir.stdout)
                            if isinstance(inspect_data, list):
                                section.running_containers = _parse_podman_inspect(inspect_data)
                        except json.JSONDecodeError:
                            pass

                # Fallback: if inspect failed, use ps data
                if not section.running_containers:
                    for c in ps_data:
                        if isinstance(c, dict):
                            section.running_containers.append(RunningContainer(
                                id=c.get("ID", ""),
                                name=c.get("Names", [""])[0] if isinstance(c.get("Names"), list) else str(c.get("Names", "")),
                                image=c.get("Image", ""),
                                status=c.get("Status", ""),
                            ))

    return section

File: cmd/inspectah/internal/inspector/container.go
// Package inspector — Container workloads inspector.
//
// Scans for Podman Quadlet unit files, docker-compose/compose YAML files,
// running containers (via podman ps + podman inspect), and installed
// Flatpak applications.
package inspector

import (
	"encoding/json"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"

	"github.com/marrusl/inspectah/cmd/inspectah/internal/schema"
)

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

// quadletExtensions lists file extensions recognised as Quadlet unit files.
var quadletExtensions = map[string]bool{
	".container": true,
	".volume":    true,
	".network":   true,
	".kube":      true,
	".pod":       true,
	".image":     true,
	".build":     true,
}

// composePatterns are glob-style base-name patterns for compose files.
var composePatterns = []string{
	"docker-compose*.yml",
	"docker-compose*.yaml",
	"compose*.yml",
	"compose*.yaml",
}

// composeSearchDirs are top-level directories to scan for compose files.
var composeSearchDirs = []string{"opt", "srv", "etc"}

// pruneMarkers and skipDirNames are defined in config.go and shared
// across inspectors for dev-artifact filtering.

const (
	nonSystemUIDMin = 1000
	nonSystemUIDMax = 60000
)

// ---------------------------------------------------------------------------
// Options
// ---------------------------------------------------------------------------

// ContainerOptions controls the Container inspector behaviour.
type ContainerOptions struct {
	QueryPodman bool
	SystemType  schema.SystemType
}

// ---------------------------------------------------------------------------
// Public entry point
// ---------------------------------------------------------------------------

// RunContainers inspects container workloads on the host: Quadlet units,
// compose files, optionally running containers, and Flatpak applications.
func RunContainers(exec Executor, opts ContainerOptions) (*schema.ContainerSection, []Warning, error) {
	section := &schema.ContainerSection{}
	var warnings []Warning

	// --- Quadlet units ---
	quadletDirs := []string{
		"/etc/containers/systemd",
		"/usr/share/containers/systemd",
		"/etc/systemd/system",
	}

	// User-level quadlets: ~/.config/containers/systemd/ for each real user.
	quadletDirs = append(quadletDirs, userQuadletDirs(exec)...)

	for _, dir := range quadletDirs {
		units := scanQuadletDir(exec, dir)
		section.QuadletUnits = append(section.QuadletUnits, units...)
	}

	// --- Compose files ---
	for _, dir := range composeSearchDirs {
		absDir := "/" + dir
		if !exec.FileExists(absDir) {
			continue
		}
		files := findComposeFiles(exec, absDir)
		section.ComposeFiles = append(section.ComposeFiles, files...)
	}

	// --- Running containers (podman) ---
	if opts.QueryPodman {
		containers, w := queryPodmanContainers(exec)
		section.RunningContainers = containers
		warnings = append(warnings, w...)
	}

	// --- Flatpak apps ---
	section.FlatpakApps = detectFlatpakApps(exec)

	return section, warnings, nil
}

// ---------------------------------------------------------------------------
// Quadlet scanning
// ---------------------------------------------------------------------------

// scanQuadletDir reads all quadlet unit files from a single directory.
func scanQuadletDir(exec Executor, dir string) []schema.QuadletUnit {
	entries, err := exec.ReadDir(dir)
	if err != nil {
		return nil
	}

	var units []schema.QuadletUnit
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		ext := filepath.Ext(entry.Name())
		if !quadletExtensions[ext] {
			continue
		}

		path := filepath.Join(dir, entry.Name())
		content, err := exec.ReadFile(path)
		if err != nil {
			continue
		}

		imageRef := ""
		if ext == ".container" {
			imageRef = extractQuadletImage(content)
		}

		// Store path relative to host root.
		relPath := strings.TrimPrefix(path, exec.HostRoot())
		relPath = strings.TrimPrefix(relPath, "/")

		units = append(units, schema.QuadletUnit{
			Path:    relPath,
			Name:    entry.Name(),
			Content: content,
			Image:   imageRef,
		})
	}
	return units
}

// extractQuadletImage parses Image= from a .container quadlet file.
func extractQuadletImage(content string) string {
	for _, line := range strings.Split(content, "\n") {
		trimmed := strings.TrimSpace(line)
		lower := strings.ToLower(trimmed)
		if !strings.HasPrefix(lower, "image") || !strings.Contains(trimmed, "=") {
			continue
		}
		key, val, ok := strings.Cut(trimmed, "=")
		if !ok {
			continue
		}
		if strings.ToLower(strings.TrimSpace(key)) == "image" {
			return strings.TrimSpace(val)
		}
	}
	return ""
}

// userQuadletDirs discovers per-user quadlet directories by parsing
// /etc/passwd for non-system UIDs (1000–59999).
func userQuadletDirs(exec Executor) []string {
	passwd, err := exec.ReadFile("/etc/passwd")
	if err != nil {
		return nil
	}

	var dirs []string
	for _, line := range strings.Split(passwd, "\n") {
		parts := strings.Split(line, ":")
		if len(parts) < 7 {
			continue
		}
		uid, err := strconv.Atoi(parts[2])
		if err != nil {
			continue
		}
		if uid < nonSystemUIDMin || uid >= nonSystemUIDMax {
			continue
		}
		home := strings.TrimPrefix(parts[5], "/")
		dirs = append(dirs, "/"+home+"/.config/containers/systemd")
	}
	return dirs
}

// ---------------------------------------------------------------------------
// Compose file discovery
// ---------------------------------------------------------------------------

// findComposeFiles recursively searches a directory for compose files,
// pruning VCS checkouts and dev-artifact directories.
func findComposeFiles(exec Executor, root string) []schema.ComposeFile {
	var matches []string
	filteredWalk(exec, root, func(path string, name string) {
		for _, pattern := range composePatterns {
			if matchGlob(pattern, name) {
				matches = append(matches, path)
			}
		}
	})
	sort.Strings(matches)

	var files []schema.ComposeFile
	for _, path := range matches {
		content, err := exec.ReadFile(path)
		if err != nil {
			continue
		}
		images := extractComposeImages(content)

		relPath := strings.TrimPrefix(path, exec.HostRoot())
		relPath = strings.TrimPrefix(relPath, "/")

		files = append(files, schema.ComposeFile{
			Path:   relPath,
			Images: images,
		})
	}
	return files
}

// filteredWalk performs a recursive directory traversal that prunes VCS
// checkouts and dev-artifact directories. For each regular file visited,
// it calls fn(fullPath, baseName).
func filteredWalk(exec Executor, dir string, fn func(path, name string)) {
	entries, err := exec.ReadDir(dir)
	if err != nil {
		return
	}

	// Check for prune markers among children.
	childNames := make(map[string]bool, len(entries))
	for _, e := range entries {
		childNames[e.Name()] = true
	}
	for marker := range pruneMarkers {
		if childNames[marker] {
			return // VCS root — skip entire subtree
		}
	}

	for _, entry := range entries {
		name := entry.Name()
		fullPath := filepath.Join(dir, name)

		if entry.IsDir() {
			if !skipDirNames[name] {
				filteredWalk(exec, fullPath, fn)
			}
			continue
		}
		fn(fullPath, name)
	}
}

// matchGlob is defined in services.go and shared across inspectors.

// extractComposeImages parses image: fields from a compose YAML without
// requiring a YAML library. Detects service-level indent dynamically so
// 2-space, 4-space, and tab-indented files all work.
func extractComposeImages(content string) []schema.ComposeService {
	var results []schema.ComposeService
	lines := strings.Split(content, "\n")

	inServices := false
	currentService := ""
	serviceIndent := -1 // indent of first service key seen

	imageRe := regexp.MustCompile(`^image:\s*(.+)`)

	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}

		indent := len(line) - len(strings.TrimLeft(line, " \t"))

		// Detect "services:" block.
		if trimmed == "services:" || strings.HasPrefix(trimmed, "services:") {
			inServices = true
			serviceIndent = -1
			currentService = ""
			continue
		}

		if !inServices {
			continue
		}

		// Top-level key other than services — stop.
		if indent == 0 && !strings.HasPrefix(trimmed, "#") {
			inServices = false
			currentService = ""
			serviceIndent = -1
			continue
		}

		// Calibrate service indent from the first indented key.
		if serviceIndent < 0 && indent > 0 {
			serviceIndent = indent
		}

		// Service-level key (e.g. "web:", "db:").
		if serviceIndent > 0 && indent == serviceIndent && strings.HasSuffix(trimmed, ":") {
			currentService = strings.TrimSuffix(trimmed, ":")
			continue
		}

		// Inside a service — look for image:.
		if currentService != "" && indent > serviceIndent {
			m := imageRe.FindStringSubmatch(trimmed)
			if m != nil {
				img := strings.Trim(m[1], "'\"")
				results = append(results, schema.ComposeService{
					Service: currentService,
					Image:   img,
				})
			}
		}
	}
	return results
}

// ---------------------------------------------------------------------------
// Podman container query
// ---------------------------------------------------------------------------

// queryPodmanContainers runs podman ps + podman inspect and returns parsed
// container data, plus any warnings.
func queryPodmanContainers(exec Executor) ([]schema.RunningContainer, []Warning) {
	var warnings []Warning

	result := exec.Run("podman", "ps", "-a", "--format", "json")
	if result.ExitCode != 0 {
		warnings = append(warnings, makeWarning(
			"containers",
			"--query-podman requested but podman ps failed -- live container data unavailable.",
		))
		return nil, warnings
	}

	stdout := strings.TrimSpace(result.Stdout)
	if stdout == "" {
		return nil, warnings
	}

	var psData []map[string]interface{}
	if err := json.Unmarshal([]byte(stdout), &psData); err != nil {
		return nil, warnings
	}

	// Collect container IDs for podman inspect.
	var ids []string
	for _, c := range psData {
		if id, ok := c["ID"].(string); ok && id != "" {
			ids = append(ids, id)
		}
	}

	// Try podman inspect for rich data.
	if len(ids) > 0 {
		args := append([]string{"inspect"}, ids...)
		ir := exec.Run("podman", args...)
		if ir.ExitCode == 0 && strings.TrimSpace(ir.Stdout) != "" {
			var inspectData []map[string]interface{}
			if err := json.Unmarshal([]byte(ir.Stdout), &inspectData); err == nil {
				containers := parsePodmanInspect(inspectData)
				if len(containers) > 0 {
					return containers, warnings
				}
			}
		}
	}

	// Fallback: parse ps data directly.
	return parsePodmanPS(psData), warnings
}

// parsePodmanInspect converts podman inspect JSON into RunningContainer
// structs.
func parsePodmanInspect(data []map[string]interface{}) []schema.RunningContainer {
	var containers []schema.RunningContainer

	for _, c := range data {
		id := stringField(c, "ID", "Id")
		name := stringField(c, "Name")
		name = strings.TrimPrefix(name, "/")

		image := stringField(c, "ImageName", "Image")
		imageID := stringField(c, "ImageID")

		state, _ := c["State"].(map[string]interface{})
		status := stringField(state, "Status")

		mounts := parseMounts(c["Mounts"])
		networks, ports := parseNetworking(c["NetworkSettings"])

		config, _ := c["Config"].(map[string]interface{})
		env := parseStringSlice(config, "Env")

		containers = append(containers, schema.RunningContainer{
			ID:       id,
			Name:     name,
			Image:    image,
			ImageID:  imageID,
			Status:   status,
			Mounts:   mounts,
			Networks: networks,
			Ports:    ports,
			Env:      env,
		})
	}
	return containers
}

// parsePodmanPS is a fallback parser when podman inspect is unavailable.
// It extracts basic fields from podman ps JSON.
func parsePodmanPS(data []map[string]interface{}) []schema.RunningContainer {
	var containers []schema.RunningContainer
	for _, c := range data {
		id := stringField(c, "ID", "Id")

		name := ""
		switch v := c["Names"].(type) {
		case []interface{}:
			if len(v) > 0 {
				name, _ = v[0].(string)
			}
		case string:
			name = v
		}

		state, _ := c["State"].(map[string]interface{})
		status := stringField(state, "Status")
		if status == "" {
			status = stringField(c, "Status", "State")
		}

		containers = append(containers, schema.RunningContainer{
			ID:     id,
			Name:   name,
			Image:  stringField(c, "Image"),
			Status: status,
		})
	}
	return containers
}

// ---------------------------------------------------------------------------
// Flatpak detection
// ---------------------------------------------------------------------------

// detectFlatpakApps lists installed Flatpak applications.
func detectFlatpakApps(exec Executor) []schema.FlatpakApp {
	// Check if flatpak is installed.
	which := exec.Run("which", "flatpak")
	if which.ExitCode != 0 {
		return nil
	}

	result := exec.Run("flatpak", "list", "--app", "--columns=application,origin,branch")
	if result.ExitCode != 0 {
		return nil
	}

	var apps []schema.FlatpakApp
	for _, line := range strings.Split(result.Stdout, "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		parts := strings.Split(line, "\t")
		if len(parts) < 2 {
			continue
		}

		branch := ""
		if len(parts) >= 3 {
			branch = strings.TrimSpace(parts[2])
		}

		apps = append(apps, schema.FlatpakApp{
			AppID:  strings.TrimSpace(parts[0]),
			Origin: strings.TrimSpace(parts[1]),
			Branch: branch,
		})
	}
	return apps
}

// ---------------------------------------------------------------------------
// JSON helpers
// ---------------------------------------------------------------------------

// stringField extracts the first non-empty string value from a map by
// trying each key in order.
func stringField(m map[string]interface{}, keys ...string) string {
	if m == nil {
		return ""
	}
	for _, k := range keys {
		if v, ok := m[k].(string); ok && v != "" {
			return v
		}
	}
	return ""
}

// parseMounts converts a Mounts JSON array into ContainerMount slices.
func parseMounts(v interface{}) []schema.ContainerMount {
	arr, ok := v.([]interface{})
	if !ok {
		return nil
	}
	var mounts []schema.ContainerMount
	for _, item := range arr {
		m, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		rw := true
		if b, ok := m["RW"].(bool); ok {
			rw = b
		}
		mounts = append(mounts, schema.ContainerMount{
			Type:        stringField(m, "Type"),
			Source:      stringField(m, "Source"),
			Destination: stringField(m, "Destination"),
			Mode:        stringField(m, "Mode"),
			RW:          rw,
		})
	}
	return mounts
}

// parseNetworking extracts networks and ports from NetworkSettings.
func parseNetworking(v interface{}) (map[string]interface{}, map[string]interface{}) {
	ns, ok := v.(map[string]interface{})
	if !ok {
		return nil, nil
	}
	networks, _ := ns["Networks"].(map[string]interface{})
	ports, _ := ns["Ports"].(map[string]interface{})
	return networks, ports
}

// parseStringSlice extracts a []string from a JSON array field.
func parseStringSlice(m map[string]interface{}, key string) []string {
	if m == nil {
		return nil
	}
	arr, ok := m[key].([]interface{})
	if !ok {
		return nil
	}
	var result []string
	for _, v := range arr {
		if s, ok := v.(string); ok {
			result = append(result, s)
		}
	}
	return result
}

